sgx/demo/src/node/tls_proxy.c

```diff
@@ -0,0 +1,658 @@
+                       key_resp->hdr.msg_size);
+
+  // send newly created 'Key Response' message
+  ret = demo_ecdh_send_msg(ecdh_svr->session.session_socket_fd, key_resp);
+  if (ret != EXIT_SUCCESS)
+  {
+    kmyth_log(LOG_ERR, "failed to send 'Key Response' message");
+    return EXIT_FAILURE;
+  }
+  kmyth_log(LOG_DEBUG, "sent 'Key Response' message");
+
+  return EXIT_SUCCESS;
+}
+
+/*****************************************************************************
+ * proxy_cleanup_defunct()
+ ****************************************************************************/
+static void proxy_cleanup_defunct()
+{
+  /* Clean up all defunct child processes. */
+  while (waitpid(-1, NULL, WNOHANG) > 0);
+}
+
+/*****************************************************************************
+ * proxy_handle_session()
+ ****************************************************************************/
+static void proxy_handle_session(TLSProxy * proxy)
+{
+  struct pollfd pfds[NUM_POLL_FDS];
+
+  int bytes_read = 0;
+  int bytes_written = 0;
+
+  unsigned char tls_msg_buf[KMYTH_TLS_MAX_MSG_SIZE];
+  unsigned char ecdh_msg_buf[KMYTH_ECDH_MAX_MSG_SIZE];
+
+  size_t ecdh_msg_len = 0;
+  ECDHPeer *ecdh_svr = &(proxy->ecdhconn);
+  BIO *tls_bio = proxy->tlsconn.bio;
+
+  secure_memset(pfds, 0, sizeof(pfds));
+  secure_memset(tls_msg_buf, 0, sizeof(tls_msg_buf));
+
+  pfds[0].fd = ecdh_svr->session.session_socket_fd;
+  pfds[0].events = POLLIN;
+
+  pfds[1].fd = BIO_get_fd(tls_bio, NULL);
+  pfds[1].events = POLLIN;
+
+  // wait to receive data with no timeout
+  // (Note: we expect to receive a 'Client Hello' message on the ECDH
+  //        interface, but no interaction should be initiated from the
+  //        KMIP server on the TLS interface. Nonetheless, we monitor
+  //        both interfaces.)
+  poll(pfds, NUM_POLL_FDS, -1);
+
+  if (pfds[0].revents & POLLIN)
+  {
+    kmyth_log(LOG_DEBUG, "ECDH receive event initiates session setup");
+
+    // execute session setup (e.g., key agreement) protocol phase
+    if (EXIT_SUCCESS == proxy_setup_ecdh_session(proxy))
+    {
+      // obtain key retrieval request from client-side of ECDH session
+      if (EXIT_SUCCESS == proxy_get_client_key_request(proxy))
+      {
+        // pass KMIP request to / receive KMIP response from key server over TLS
+        if (EXIT_SUCCESS == proxy_get_kmip_response(proxy))
+        {
+          // return 'retrieve key' response to the client that submitted request
+          if (EXIT_SUCCESS != proxy_send_key_response_message(proxy))
+          {
+            kmyth_log(LOG_DEBUG, "failed to send 'Key Response' message");
+          }
+        }
+        else
+        {
+          kmyth_log(LOG_DEBUG, "failed to retrieve KMIP 'get key' response");
+        }
+      }
+      else
+      {
+        kmyth_log(LOG_DEBUG, "failed to receive 'Key Request' message");
+      }
+    }
+    else
+    {
+      kmyth_log(LOG_DEBUG, "failed to setup ECDH session (with client)");
+    }
```

Comment:
I believe that the original TLS proxy design was more versatile. It setup a session key for use between the enclave client and the proxy using an ECDH key agreement. It likewise setup a secure connection between the proxy and the server over TLS. After this setup, the proxy could then securely pass any set of messages between the enclave client and the server, in either direction, On its interface with the client, it encrypted messages using the session key derived using ECDH and on the interface with the server, it had a TLS connection.

The new TLS proxy is a participant in a 'retrieve key' protocol involving a client (intended to be in an SGX enclave), a proxy, and a server. While this limits the potential for direct application of the proxy for other use cases, I believe that this change is appropriate for implementing this new protocol.
- Currently the proposed protocol is specified to perform a single request/response exchange with the server per session. Modifications could potentially be considered to handle multiple or arbitrary requests/responses in a single ECDH session, but I don't think that is a necessary consideration for this demo.
- Currently the protocol is specified to use KMIP formatted request/response payloads. I believe the proxy will forward arbitrarily formatted request/response message pairs, but the client code would have to be modified to address a server that used a different request/response format. Again, I think this is OK for now.
- Even though each key retrieval requires its own ECDH session, the ECDH session limit as a configurable parameter (that defaults to unlimited) and supports testing that requires multiple or an arbitrary number of 'retrieve key' events.
- Yes, this code moves the message encryption and decryption functionality into the protocol code (message compose and parse functions). While this could be returned to the communication layer, I'm not sure I understand why that would be an improvement.